### DISCUSSION TITLE
Get sending provider out of receipt

### DISCUSSION THREAD

**Discussion by PPotappel** [OxalisCommunity/oxalis#642] [Q&A]
Hi,

We try to collect data for the Peppol Transactions Reporting.
For outgoing messages we can get the CertSubjectCN out of the public key in the MDN in the RECEIPT.
But for **incoming** messages the public key is lacking in the MDN.
How can we get to know the SP-id of the Sender?

We are using Oxalis v5.

**Comment by AlphonseSantoro** [comment]
@PPotappel Did you find an anwer to this? I've tried to do lookups based on the sender participant, but that is error prone due to the participant could no longer be registered. I think a solution will be to log this as we receive it, but is there anyway to get this without the logs?

**Comment by dladlk** [comment]
On receiving, Oxalis default persister saves receipts in inboundFolder subfolders as %transmissionIdentidier%.receipt.dat. If default evidence factory is used, it is an XML where you can find SPID in SenderDetails AttributedElectronicAddress like this:

```
	<EventTime>2024-02-16T13:34:36+02:00</EventTime>
	<SenderDetails>
		<AttributedElectronicAddress scheme="iso6523-actorid-upis">pno000063</AttributedElectronicAddress>
	</SenderDetails>
	<RecipientsDetails>
		<EntityDetails>
			<AttributedElectronicAddress scheme="iso6523-actorid-upis">pdk000253</AttributedElectronicAddress>
		</EntityDetails>
	</RecipientsDetails>
```
So you do not need to extract it from sender certificate on receiving - as it is filled by sender in SOAPHeaders and SBD and passed to receipt evidence (and actually it really misses the certificate of sender).

**Reply by AlphonseSantoro** [reply]
That is not the case in our receips, the `SenderDetails` and the `RecipientsDetails` contain the participant id not the SPID.

**Reply by dladlk** [reply]
You are right... 

Original UserMessage/From is there only if you set headerParser to non-default SbdhHeaderParser (e.g. by disabling header module) - as in this case it will fill sender/receiver with SOAP UserMessage and not with SBDH:

https://github.com/OxalisCommunity/Oxalis-AS4/blob/db0ec7003b16f5e112c5d7b62aea3fe0ead4c81e/src/main/java/network/oxalis/as4/inbound/As4InboundHandler.java#L361-L371

**Reply by wartab** [reply]
@dladlk Can you indicate concrete steps to obtain the sender details and receiver details when respectively receiving and sending documents? We are using the .war files with Tomcat to receive documents and the standalone .jar to send them. 
There is not much info out there regarding this, but every small bit of information I find leads to having to somehow implement a plugin for Oxalis or modify Oxalis in some way. 
I can't seem to be able to find any configuration option to get a hold of these details, despite them being mandatory since 2024 for statistics reports.

**Reply by dladlk** [reply]
Can you attach here 2 examples of your receipt.dat: one which you have when you send and one when you receive?

When you send (C2 side) - I think that base64 encoded OriginalReceipt in Extension actually contains base64 encoded receiver certificate, as it contains the whole Envelope with Messaging/SignalMessage response INCLUDING wsse:Security tag. So you can extract SPID of receiver at least there if it is not logged anywhere else.

When you receive (C3 side) - I am quite surprised that I cannot see anything related to sender certificate in OriginalReceipt - as its Envelope does not include wsse:Security tag (because it is not yet generated yet, I guess) - so it misses any certificate...

If it is so, I see no other way without changing the code then to disable SBD header parsing, so SOAP Envelope from/to is not overridden by originalSender/finalReceipient identifiers. But in this case your documents will be saved in folders with another names - not like 0088_1234567890123\0088_3210987654321 (where 1234567890123 and 3210987654321 are GLNs of receiver and sender from SBDH), but like
pno000123\pse000124 - where first folder should be your own SPID, second - sender SPID...

To disable SBD header parsing, try to add to your oxalis.conf
```
oxalis.header.parser = dummy
```

In logs you should see:
```
Header => PARSER: dummy
```
And in this case, your receipt.dat for receiving will contain SenderDetails/RecipientsDetails with pno000123/pse000124 instead of original identifiers, so you can use it for reporting after converting to upper case.

But this solution depends on how do you load documents further - maybe you need them to be placed into different folders for different receiver numbers...

**Reply by aaron-kumar** [reply]
@PPotappel , @AlphonseSantoro , @wartab : Please refer : https://github.com/OxalisCommunity/oxalis/wiki/Peppol-Reporting
